closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -474,12 +576,273 @@ JNIEXPORT jlong JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
+    }
+
+    return ossl_ver;
+}
+
+static void *
+load_crypto_library(jboolean traceEnabled, const char *libName)
+{
+    void *result = NULL;
+    if ((NULL != libName) && ('\0' != *libName)) {
+#if defined(_AIX)
+        int flags = RTLD_NOW;
+        if (NULL != strrchr(libName, '(')) {
+            flags |= RTLD_MEMBER;
+        }
+        result = dlopen(libName, flags);
+#elif defined(_WIN32) /* defined(_AIX) */
+        result = LoadLibrary(libName);
+#else /* defined(_WIN32) */
+        result = dlopen(libName, RTLD_NOW);
+#endif /* defined(_AIX) */
+    }
+    return result;
+}
+
+/* Look for a crypto library in java.home or the system.
+ * NULL is returned when an appropriate crypto library
+ * cannot be found.
+ */
+static void *
+find_crypto_library(jboolean traceEnabled, const char *chomepath)
+{
+    void *result = NULL;
+    void *prevResult = NULL;
+    size_t numOfLibs = 0;
+    size_t i = 0;
+    long tempVersion = 0;
+    long previousVersion = 0;
+    size_t num_of_generic = 0;
+
+#if defined(_AIX)
+    num_of_generic = 4;
+#elif defined(__linux__) /* defined(_AIX) */
+    num_of_generic = 1;
+#endif /* defined(_AIX) */
+
+    /* Library names for OpenSSL 1.1.1, 1.1.0 and symbolic links:
+     * It is important to preserve the order!!!
+     *
+     * Since there is no indication of the version of a symlink,
+     * they have to be loaded first, so as to compare with other
+     * available options.
+     *
+     * The rest of the libraries are listed in descending order,
+     * which allows us to do two things:
+     * - Stop if a general symlink is loaded and we then find a
+     *   specific version that is higher.
+     * - Stop immediately if a specific version is loaded, as
+     *   anything after that will be a lower version.
+     */
+    static const char * const libNames[] = {
+#if defined(_AIX)
+        "libcrypto.a(libcrypto64.so)",      /* general symlink library name from archive file */
+        "libcrypto64.so",                   /* general symlink library name */
+        "libcrypto.a(libcrypto.so)",        /* general symlink library name from archive file */
+        "libcrypto.so",                     /* general symlink library name */
+        "libcrypto.a(libcrypto64.so.3)",    /* 3.x library name from archive file */
+        "libcrypto64.so.3",                 /* 3.x library name */
+        "libcrypto.a(libcrypto.so.3)",      /* 3.x library name from archive file */
+        "libcrypto.so.3",                   /* 3.x library name */
+        "libcrypto.a(libcrypto64.so.1.1)",  /* 1.1.x library name from archive file */
+        "libcrypto.so.1.1",                 /* 1.1.x library name */
+        "libcrypto.so.1.0.0",               /* 1.0.x library name */
+#elif defined(__APPLE__) /* defined(_AIX) */
+        /* "libcrypto.dylib"                In MacOS 11 or later, loading this symlink causes a
+                                            fatal warning and associated abort by default. */
```

Comment:
That's better, but the `*/` of a multi-line comment should be on a line by itself.
Please correct the spelling of "macOS", and I think It should say "on" (a platform) rather than "in".
Another approach would be to mention the exception for macOS when explaining why symlinks are listed first.

Reply:
You mean in the overall comment?

Reply:
Added wording as a note in the comment above the libraries.